src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs

```diff
@@ -88,39 +88,38 @@ internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonConverter<Ke
             return new KeyValuePair<TKey, TValue>(k, v);
         }
 
-        private T ReadProperty<T>(ref Utf8JsonReader reader, JsonSerializerOptions options)
+        private T ReadProperty<T>(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
             T k;
-            Type typeToConvert = typeof(T);
 
-            // Attempt to use existing converter first before re-entering through JsonSerializer.Read().
-            JsonConverter<T> keyConverter = options.GetConverter(typeToConvert) as JsonConverter<T>;
-            if (keyConverter == null)
+            // Attempt to use existing converter first before re-entering through JsonSerializer.Deserialize().
+            if (typeToConvert != typeof(object) && (options.GetConverter(typeToConvert) is JsonConverter<T> keyConverter))
```

Comment:
Add a comment here please; something like "the default converter for object does not support writing"

Reply:
Also this is a case for a future feature for having a way for converters to say they just support reading (or writing). That would allow us to use the normal (non-`System.Object`) converters to avoid having to do this check here.